### PR TITLE
Indirect data migration for crane transfer-pvc

### PR DESCRIPTION
**Pull request by aufi**
## Summary

Proposes extending `crane transfer-pvc` with a `--cloud-storage` flag that enables indirect PVC data migration via S3-compatible cloud storage.

Instead of direct rsync-over-stunnel between clusters, data flows through an intermediary bucket: source cluster uploads via rclone, destination cluster downloads.

  This unlocks three scenarios that the current direct transfer cannot handle:
  - Cross-cloud migration (AWS → GCP) without direct cluster connectivity
  - Air-gapped / network-restricted environments sharing access to internal MinIO/S3
  - Simplified network setup (no Ingress/Route/stunnel, just outbound HTTPS)

  ## Key Design Decisions

  - **rclone as Go library** compiled into the existing `migtools/rsync-transfer` image
  - **No new CLI commands** — single `--cloud-storage` flag changes the transfer mechanism transparently; existing behavior is 100% preserved
  - **Credentials** via K8s Secret (`--rclone-config-secret`) or local file (`--rclone-config-file`, crane creates temporary Secrets automatically)

## Summary by CodeRabbit

* **Documentation**
  * Added documentation for an optional indirect cloud-storage transfer mode for PVC migrations using an S3-compatible workflow.
  * Described CLI flags for cloud storage, rclone credentials and configuration, optional client-side encryption, and cloud cleanup controls.
  * Clarified transfer behavior, validation rules, metadata and symlink handling, security considerations, and upgrade or downgrade guidance.
  * Included testing guidance for unit, integration, and end-to-end coverage with an S3-compatible backend.

### CODE REVIEW

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+
+The `migtools/rsync-transfer` image gains rclone as a Go dependency. To
+minimize image size growth, only the required backends are imported:
+
+```go
+import (
+    _ "github.com/rclone/rclone/backend/s3"
+    _ "github.com/rclone/rclone/backend/local"
+    _ "github.com/rclone/rclone/backend/crypt"    // optional: --encrypt
+)
```
[the remark this answers is not in the document]
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md:217]: re-check, I dont see the issue

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+2. **Source phase:** Create mover Pod in source cluster
+   - Image: `quay.io/konveyor/rsync-transfer:latest`
+   - Mount source PVC as ReadOnly at `/mnt/pvc-data`
+   - Mount rclone config Secret at `/etc/rclone/rclone.conf`
+   - Run: `rclone sync /mnt/pvc-data <cloud-storage-path>`
+   - Follow Pod logs, garbage collect on completion
+3. **Destination phase:** Create mover Pod in destination cluster
+   - Mount destination PVC as ReadWrite at `/mnt/pvc-data`
+   - Mount rclone config Secret
+   - Run: `rclone sync <cloud-storage-path> /mnt/pvc-data`
```
[the remark this answers is not in the document]
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md]: documented.

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+When `--cloud-storage` is provided, `crane transfer-pvc` replaces the
+rsync/stunnel flow with two sequential rclone operations:
+
+1. **Upload phase (source cluster):** A mover Pod mounts the source PVC
+   read-only and uses rclone to sync its contents to the specified cloud
+   storage location.
+
+2. **Download phase (destination cluster):** A mover Pod mounts the
+   destination PVC read-write and uses rclone to sync the contents from
+   cloud storage into the PVC.
```
[the remark this answers is not in the document]
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md:109]: right, mentioned in document below, re-check

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+
+\* One of `--rclone-config-secret` or `--rclone-config-file` is required
+when `--cloud-storage` is set.
+
+**Validation rules:**
+
+- `--cloud-storage` without rclone config → error
+- Both `--rclone-config-secret` and `--rclone-config-file` at once → error
+- Referenced Secret does not exist in cluster → error before creating mover Pod
+
```
[the remark this answers is not in the document]
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md:201]: not part of the enhancement, will work on most suitable solution during implementation

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+The `migtools/rsync-transfer` image gains rclone as a Go dependency. To
+minimize image size growth, only the required backends are imported:
+
+```go
+import (
+    _ "github.com/rclone/rclone/backend/s3"
+    _ "github.com/rclone/rclone/backend/local"
+)
+```
```
[the remark this answers is not in the document]
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md:218]: it is example

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+- Runs the rclone sync command inside the Pod
+- Streams Pod logs for progress reporting
+- Removes the Pod on completion (garbage collection)
+
+#### Transfer Flow
+
+1. Crane CLI validates flags
+2. **Source phase:** Create mover Pod in source cluster
+   - Image: `quay.io/konveyor/rsync-transfer:latest`
+   - Mount source PVC as ReadOnly at `/mnt/pvc-data`
```
[the remark this answers is not in the document]
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md:239]: it is example, upstream main development uses latest

File: enhancements/crane-2.1/indirect-data-migration/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md:109]: again, solved in orchestration layer (and docs until having it)

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 7]
```diff
@@ -0,0 +1,464 @@
...
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] User-facing documentation is created
+
+## Open Questions
+
+1. Should the `--cloud-storage` flag accept non-S3 rclone remotes (GCS, Azure) in the first iteration, or limit to S3-compatible only?
```
Comment (rromannissen) [inline enhancements/crane-2.1/indirect-data-migration/README.md]: I think we can start with S3 first, and then add additional support in subsequent iterations.

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 8]
```diff
@@ -0,0 +1,464 @@
...
+2. Should crane automatically clean up data from cloud storage after a successful transfer, or leave that to the user?
```
Comment (rromannissen) [inline enhancements/crane-2.1/indirect-data-migration/README.md]: What about having a specific flag for that? I can imagine defaulting to cleanup, and having a flag --keep-cloud-data to leave it there. Nevertheless, can I ask about the specific use case for something like that?

File: enhancements/crane-2.1/indirect-data-migration/README.md [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+3. Should `--bandwidth-limit` be included in the first release or deferred to a follow-up?
```
Comment (rromannissen) [inline enhancements/crane-2.1/indirect-data-migration/README.md]: Could you elaborate further on what this would mean and imply? What are the consequences of not having it, and what are the advantages of including it?

File: enhancements/crane-2.1/indirect-data-migration/README.md (hunk 8 above)
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md]: +1 on cleanup by default; allow users to keep the transfer storage data was more for possible debugging or some _kind of_ backup.

File: enhancements/crane-2.1/indirect-data-migration/README.md (hunk 9 above)
Comment (aufi) [reply enhancements/crane-2.1/indirect-data-migration/README.md]: The `--bandwith-limit` could allow limit bandwith of the data transfer to make it more bearable by the cluster (primary in overload and similar edge-cases)

The question with the `--bandwith-limit` exact flag is probably too detailed, we have an issue (actually created after this enhancement) on testing&verifying data transfer sync flags usage https://github.com/migtools/crane/issues/689, so we'll identify just really impactful flags, so we might discuss it later if needed.

For now, I'd replace this question with note, that for 0.11 release, we will not add additional flags unless directly requested or found critical having it during testing.

**Comment by istein1** [comment]
## Review Comments

### Substantive design/safety concerns

**1. `rclone sync` deletes data on the destination — needs a prominent warning**

The doc states: *"Files in the target that are absent from the source will be deleted — any existing data on the cloud storage path or the destination PVC will be overwritten."* This is buried in the transfer flow section. `sync` is destructive to the destination — if a user accidentally points `--cloud-storage` at a bucket path with existing data, or if the destination PVC has data they want to keep, it gets wiped. This should be:
- Called out in the CLI flag table for `--cloud-storage` (e.g., "Warning: overwrites existing data at this path")
- Mentioned in the docs team recommendations for a warning box in user-facing guides
- Considered whether `rclone copy` (additive, no deletes) should be the default, with `sync` as opt-in via a `--sync-mode` flag

**2. Metadata round-trip fidelity through S3 is unclear**

The doc says *"Filesystem metadata (mode bits, ownership, xattrs) is preserved using rclone's `--metadata` flag."* But S3 doesn't natively store POSIX metadata — rclone uses object metadata headers for this, and support varies by provider and rclone version. For Kubernetes workloads, UID/GID preservation is critical (especially on OpenShift with namespace UID ranges). The test plan should explicitly include a metadata fidelity test: create files with specific ownership/permissions on the source PVC, transfer through S3, and verify they match on the destination. If metadata doesn't survive the round-trip reliably, this needs to be documented as a limitation.

**3. Single rclone.conf may not work for cross-cloud transfers**

Story 1 is "AWS EKS to GCP GKE" — but the rclone config Secret must exist in both clusters with the same content. If the S3 bucket is in AWS, the GCP cluster needs AWS credentials. That works, but what if the intermediary storage is in neither cloud and each cluster needs different credentials to reach it? Or if the source cluster uses IAM roles (no static keys) while the destination uses static credentials? Consider whether `--rclone-config-secret` should accept two values (source and destination), or whether two separate flags (`--rclone-config-secret-source`, `--rclone-config-secret-destination`) are needed for asymmetric credential setups. This may be a v2 concern, but it directly affects the cross-cloud story.

**4. Orphaned Secret garbage collection needs detail**

The doc says orphaned Secrets from `--rclone-config-file` crashes are cleaned up "on next run." How does crane identify which Secrets are orphaned? Are they labeled with a crane-specific label? If not, crane would need to guess — or worse, leave them indefinitely. The labeling/naming scheme for temporary Secrets should be specified so the GC mechanism is concrete.

### Minor gaps

**5. `--compress` mentioned in PR body but not in the enhancement**

The PR summary lists `--compress` as an optional flag, but it doesn't appear in the CLI flag table or anywhere in the document body. Should it be added, or was it cut?

**6. Test plan doesn't cover symlink edge cases**

The doc mentions `.rclonelink` handling but the test plan doesn't include symlinks, absolute symlinks, or symlink loops as test scenarios.

**Comment by aufi** [comment]
Thanks for review, comment below.

> ### Substantive design/safety concerns
> **1. `rclone sync` deletes data on the destination — needs a prominent warning**
> 
> The doc states: _"Files in the target that are absent from the source will be deleted — any existing data on the cloud storage path or the destination PVC will be overwritten."_ This is buried in the transfer flow section. `sync` is destructive to the destination — if a user accidentally points `--cloud-storage` at a bucket path with existing data, or if the destination PVC has data they want to keep, it gets wiped. This should be:
> 
> * Called out in the CLI flag table for `--cloud-storage` (e.g., "Warning: overwrites existing data at this path")
> * Mentioned in the docs team recommendations for a warning box in user-facing guides
> * Considered whether `rclone copy` (additive, no deletes) should be the default, with `sync` as opt-in via a `--sync-mode` flag

Added a warning to table with `--cloud-storage` flag, but it looks native&expected to me, it might be part of docs, but looks enough mentioned in the enhancement now.
 
> **2. Metadata round-trip fidelity through S3 is unclear**
> 
> The doc says _"Filesystem metadata (mode bits, ownership, xattrs) is preserved using rclone's `--metadata` flag."_ But S3 doesn't natively store POSIX metadata — rclone uses object metadata headers for this, and support varies by provider and rclone version. For Kubernetes workloads, UID/GID preservation is critical (especially on OpenShift with namespace UID ranges). The test plan should explicitly include a metadata fidelity test: create files with specific ownership/permissions on the source PVC, transfer through S3, and verify they match on the destination. If metadata doesn't survive the round-trip reliably, this needs to be documented as a limitation.

Good point, added note to test this with option to either document limitations or fallback to tar format (but this might bring unwanted sideeffects). 

> **3. Single rclone.conf may not work for cross-cloud transfers**
> 
> Story 1 is "AWS EKS to GCP GKE" — but the rclone config Secret must exist in both clusters with the same content. If the S3 bucket is in AWS, the GCP cluster needs AWS credentials. That works, but what if the intermediary storage is in neither cloud and each cluster needs different credentials to reach it? Or if the source cluster uses IAM roles (no static keys) while the destination uses static credentials? Consider whether `--rclone-config-secret` should accept two values (source and destination), or whether two separate flags (`--rclone-config-secret-source`, `--rclone-config-secret-destination`) are needed for asymmetric credential setups. This may be a v2 concern, but it directly affects the cross-cloud story.

The S3 needs to be reached from both sides, if cloud providers would require different access keys from in-cluster and from internet connections, we might just document it as a limitation for this release. Added to known limitations section of the enhancement.

> **4. Orphaned Secret garbage collection needs detail**
> 
> The doc says orphaned Secrets from `--rclone-config-file` crashes are cleaned up "on next run." How does crane identify which Secrets are orphaned? Are they labeled with a crane-specific label? If not, crane would need to guess — or worse, leave them indefinitely. The labeling/naming scheme for temporary Secrets should be specified so the GC mechanism is concrete.

Yes, crane-specific labels are expected to be used for these resources. Added mention about it explicitly.

> ### Minor gaps
> **5. `--compress` mentioned in PR body but not in the enhancement**
> 
> The PR summary lists `--compress` as an optional flag, but it doesn't appear in the CLI flag table or anywhere in the document body. Should it be added, or was it cut?

Thanks, this was a leftover, compress was removed from this enhancement.

> **6. Test plan doesn't cover symlink edge cases**
> 
> The doc mentions `.rclonelink` handling but the test plan doesn't include symlinks, absolute symlinks, or symlink loops as test scenarios.

Right, test plan was updated with this point.

**Review by istein1 — APPROVED** [review]
